### ISSUE TITLE
Fastcall/vectorcall specialized argument types

### ISSUE THREAD

**Issue by da-woods** [cython/cython#3343]
I've been looking at the code for fastcall and vectorcall functions, and one issue seems to be that for a function like:

    def f(a, *args, **kwds):
        # something...

the wrapper function immediately generates a tuple for `args` and a dict for `kwds`, which presumably misses most of the benefits of the more efficient calling method.

What I was considering is creating some types `cython.fastcalltuple` and `cython.fastcalldict` that `args` and `kwds` could be typed with. These would ultimately reduce to a basic, non-refcounted structure, and so could be used directly and more efficiently. However, they would only support very limited operations.

For example `cython.fastcalltuple` could support:
* integer indexing
* very simple slicing (integer start and end, positive step)
* Assignment to split into multiple values `a, b, c = args`
* Explicit conversion to a Python sequence (e.g. `tuple(args)`)
* Maybe calling other functions with them under very limited circumstances (i.e. when they can be fed directly into a fastcall/vectorcall)

Anything else would fail at compile-time, so it couldn't auto-coerce to a Python object. `cython.fastcalldict` would be similarly restricted (although I've given it less thought).

The idea being that for a lot of functions these operations are all that's ever done to the star/starstar args, so this provides a fairly easy user-selectable optimization. (Obviously a future step might be to apply the optimization automatically when possible, but that's obvious a bit harder).

-------------------------------------------------

It's possible that I've mis-understood something and this wouldn't provide real benefits, hence I'm creating an issue in advance of doing anything so that I can be told why I'm wrong.

**Comment by scoder** [comment]
Good idea. It wouldn't have to fail at compile time, though. Instead, we could infer these types for the two arguments automatically and then have them auto-coerce to tuple and/or dict at need (and separately), when we detect unsafe usages. We do this for potentially overflowing arithmetic as well, which turns inferred C integer variables into Python object variables.

**Comment by axeldavy** [comment]
I would suggest not to type with cython.fastcalltuple and cython.fastcalldict specifically, but instead cython.callargs and cython.callkwds.

In addition to the listed operations, could also be supported 'contains' (using the optimized code written for #6684), 'pop' and iterating (`.keys()`, `.items()`, `.values()`).

All the tuple and dict operations could be supported, except the direct cpython operations. tuple() and dict() would allow the explicit conversion to be allowed those calls. Alternatively, when they are used as object somewhere in the code, they would get automatically promoted to tuple()/dict() just before (and only once).

The advantage of the change is that under the hood, cython would be allowed to duplicate the implementation, one where the vectorcall protocol is used, and one when it isn't. For item instanciation for instance, both paths could co-exist.

If vectorcall is not used, `tuple(args)` and `dict(kwargs)` would be no-op.

Finally passing kwargs should be optimized. If anything, #6684 shows that building a dict out of vectorcall parameters is not cheap. It should be avoided to promote kwargs to dict when calling a method (and the method should be called with the vectorcall protocol if possible). In addition if one wants to build a args or kwargs to pass to a method, cython.callargs and cython.callkwds should be accessible to build ones.

**Comment by da-woods** [comment]
I had an implementation, but it's significantly out of date and probably too complicated for the fairly marginal benefits it brought.

> I would suggest not to type with cython.fastcalltuple and cython.fastcalldict specifically, but instead cython.callargs and cython.callkwds.

I think my view was that it should be transparent and not something you can request - just something that Cython does for you provided you don't do anything unsupported on the arguments. On that basis the name doesn't matter - it's internal only.

> In addition to the listed operations, could also be supported 'contains' (using the optimized code written for [#6684](https://github.com/cython/cython/pull/6684)), 'pop' and iterating (`.keys()`, `.items()`, `.values()`).

I think there's a strong argument for keeping anything like this as simple as possible. At least for the pseudo-dict I don't think it's worth implementing much beyond just passing it to another function. Even that isn't easy because the keys end up tied up with the args tuple.

**Comment by axeldavy** [comment]
I would tend to disagree, and favor various pythonic fast accessors. Or if not, ability to access the internals and have users implement what they want with it.

I'm always surprised to see Cython being used in almost all well known Python libraries, and people have various uses. That's why I'd favor having items that behave like dict and tuple, but avoid creating a python item unless needed.

To give one example of 'complex' kwargs handling that benefits from a dict-like interface (here I'm citing one of my code, but there must be many other complex usages as well):
- Using pop to handle some kwargs before others, if defined 
- passing the kwargs after the pop operation to another function that takes **kwargs
- iterate through the kwargs (and doing `setattr(self, name, value)`)

Some libraries also seem to use kwargs names as user-defined keys (tables headers, etc).

That's why I believe it would be leaner and more pythonic if the arguments would be behaving like tuple/dict and the actual backing (vectorcall or actual tuple/dict) would be made such that it doesn't matter which is behind (no specific code handling on user side).

**Comment by axeldavy** [comment]
Here are some links of examples of complex kwargs usage:
[dearcygui](https://github.com/DearCyGui/DearCyGui/blob/179105b7388dd107d0684f5fc9aba3a85c872ca7/dearcygui/core.pyx#L1405)
[dearcygui](https://github.com/DearCyGui/DearCyGui/blob/179105b7388dd107d0684f5fc9aba3a85c872ca7/dearcygui/widget.pyx#L1295)
[h5py](https://github.com/h5py/h5py/blob/e4aa0d2da6082aaaf11746f3154cdd1588992cf8/h5py/_hl/files.py#L577)

**Comment by da-woods** [comment]
`pop` is definitely not an option - the sensible (i.e. copy-free) implementation of this needs to be read-only.

Iteration could work fairly straightforwardly - it's mainly just that the more you implement the more complicated the feature is. 

`contains` again could work. It'd have difficult performance characteristics than people would expect from a `dict` (probably better for small numbers of args, but would scale worse) so would be a bit of a pitfall if you applied it automatically and thus I think it'd probably be a mistake. 

-----------

It's probably largely academic for now - I don't think anyone really has the inclination to dig out and fix my old PR for this.
